File: packages/core/src/retrieval/indexer.ts
/**
 * Copyright 2026 CodeSetu Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { chunkFile, type ChunkOptions } from "./chunker.js";
import type { WorkspaceIndex } from "./store.js";
import type { CodeChunk, EmbeddingProvider, IndexedChunk, WorkspaceFile } from "./types.js";

/** How many chunk texts to send per embedding request. */
export const DEFAULT_EMBED_BATCH_SIZE = 64;

export interface IndexUpdateOptions extends ChunkOptions {
  batchSize?: number;
  /** Invoked as files are embedded, for progress reporting. */
  onProgress?: (done: number, total: number) => void;
  signal?: AbortSignal;
}

export interface IndexUpdateResult {
  /** Files that were (re)chunked and embedded this run. */
  indexed: number;
  /** Files skipped because their content hash was unchanged. */
  skipped: number;
  /** Files dropped because they no longer exist in the workspace. */
  removed: number;
}

/**
 * Fast, stable, non-cryptographic content hash (FNV-1a, 32-bit) used only to
 * detect whether a file changed since it was last indexed. No crypto dependency,
 * so the indexer stays portable across host environments.
 */
export function hashContent(text: string): string {
  let hash = 0x811c9dc5;
  for (let i = 0; i < text.length; i += 1) {
    hash ^= text.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return (hash >>> 0).toString(16);
}

/**
 * Bring `index` in line with `files`: skip unchanged files, (re)chunk and embed
 * changed/new ones, and drop files that have disappeared. Embedding is batched
 * and incremental, so steady-state re-index after editing one file is cheap.
 */
export async function updateWorkspaceIndex(
  index: WorkspaceIndex,
  provider: EmbeddingProvider,
  files: readonly WorkspaceFile[],
  options: IndexUpdateOptions = {},
): Promise<IndexUpdateResult> {
  const batchSize = Math.max(1, options.batchSize ?? DEFAULT_EMBED_BATCH_SIZE);
  const present = new Set(files.map((file) => file.path));

  let removed = 0;
  for (const path of index.paths()) {
    if (!present.has(path)) {
      index.removeFile(path);
      removed += 1;
    }
  }

  // Collect every chunk that needs (re)embedding across all changed files, so a
  // single small file doesn't waste a whole embedding round-trip.
  const pending: Array<{ path: string; hash: string; chunks: CodeChunk[] }> = [];
  let skipped = 0;
  for (const file of files) {
    const hash = hashContent(file.text);
    if (index.hasUnchanged(file.path, hash)) {
      skipped += 1;
      continue;
    }
    const chunks = chunkFile(file.path, file.text, options);
    if (chunks.length === 0) {
      // Empty/whitespace file: record the hash so it's not re-chunked next run.
      index.upsertFile(file.path, hash, []);
      continue;
    }
    pending.push({ path: file.path, hash, chunks });
  }

  const flatChunks = pending.flatMap((entry) => entry.chunks);
  const vectors = await embedAll(provider, flatChunks, batchSize, options);

  // Stitch the flat vector list back onto each file's chunks, in order.
  let cursor = 0;
  let indexed = 0;
  for (const entry of pending) {
    const indexedChunks: IndexedChunk[] = entry.chunks.map((chunk) => ({
      ...chunk,
      id: `${chunk.path}:${chunk.startLine}-${chunk.endLine}`,
      vector: vectors[cursor++]!,
    }));
    index.upsertFile(entry.path, entry.hash, indexedChunks);
    indexed += 1;
  }

  return { indexed, skipped, removed };
}

async function embedAll(
  provider: EmbeddingProvider,
  chunks: readonly CodeChunk[],
  batchSize: number,
  options: IndexUpdateOptions,
): Promise<number[][]> {
  const vectors: number[][] = [];
  for (let start = 0; start < chunks.length; start += batchSize) {
    if (options.signal?.aborted === true) {
      throw new Error("Indexing was cancelled.");
    }
    const batch = chunks.slice(start, start + batchSize);
    const embedded = await provider.embed(batch.map((chunk) => chunk.text));
    vectors.push(...embedded);
    options.onProgress?.(Math.min(start + batch.length, chunks.length), chunks.length);
  }
  return vectors;
}

File: apps/jetbrains/src/main/kotlin/ai/codesetu/retrieval/Indexer.kt
package ai.codesetu.retrieval

const val DEFAULT_CHUNK_MAX_LINES = 60
const val DEFAULT_CHUNK_OVERLAP = 10
const val DEFAULT_EMBED_BATCH_SIZE = 64
const val DEFAULT_RETRIEVAL_K = 8

/**
 * Split a file into overlapping, line-aligned chunks. Line-based keeps it
 * language-agnostic; the overlap means a function straddling a boundary still
 * lands wholly inside some chunk. Whitespace-only input produces nothing.
 */
fun chunkFile(
  path: String,
  text: String,
  maxLines: Int = DEFAULT_CHUNK_MAX_LINES,
  overlap: Int = DEFAULT_CHUNK_OVERLAP,
): List<CodeChunk> {
  if (text.isBlank()) return emptyList()
  val max = maxOf(1, maxLines)
  val ov = maxOf(0, minOf(overlap, max - 1))
  val step = max - ov
  val lines = text.split("\n")
  val chunks = ArrayList<CodeChunk>()
  var start = 0
  while (start < lines.size) {
    val end = minOf(start + max, lines.size)
    val slice = lines.subList(start, end)
    if (slice.joinToString("").isNotBlank()) {
      chunks.add(CodeChunk(path, start + 1, end, slice.joinToString("\n")))
    }
    if (end >= lines.size) break
    start += step
  }
  return chunks
}

/**
 * Fast, stable, non-cryptographic content hash (FNV-1a, 32-bit) used only to
 * detect whether a file changed since it was last indexed.
 */
fun hashContent(text: String): String {
  var hash = 0x811c9dc5.toInt()
  for (ch in text) {
    hash = hash xor ch.code
    hash *= 0x01000193
  }
  return Integer.toHexString(hash)
}

data class IndexUpdateResult(val indexed: Int, val skipped: Int, val removed: Int)

/**
 * Bring [index] in line with [files]: skip unchanged files, (re)chunk and embed
 * changed/new ones (via [embed]), and drop files that have disappeared.
 * Embedding is batched and incremental, so re-index after editing one file is
 * cheap. Kotlin mirror of the TypeScript `updateWorkspaceIndex`.
 */
fun updateWorkspaceIndex(
  index: WorkspaceIndex,
  embed: (List<String>) -> List<List<Double>>,
  files: List<WorkspaceFile>,
  maxLines: Int = DEFAULT_CHUNK_MAX_LINES,
  overlap: Int = DEFAULT_CHUNK_OVERLAP,
  batchSize: Int = DEFAULT_EMBED_BATCH_SIZE,
  onProgress: ((Int, Int) -> Unit)? = null,
): IndexUpdateResult {
  val present = files.map { it.path }.toHashSet()
  var removed = 0
  for (path in index.paths()) {
    if (path !in present) {
      index.removeFile(path)
      removed += 1
    }
  }

  data class Pending(val path: String, val hash: String, val chunks: List<CodeChunk>)
  val pending = ArrayList<Pending>()
  var skipped = 0
  for (file in files) {
    val hash = hashContent(file.text)
    if (index.hasUnchanged(file.path, hash)) {
      skipped += 1
      continue
    }
    val chunks = chunkFile(file.path, file.text, maxLines, overlap)
    if (chunks.isEmpty()) {
      // Empty/whitespace file: record the hash so it's not re-chunked next run.
      index.upsertFile(file.path, hash, emptyList())
      continue
    }
    pending.add(Pending(file.path, hash, chunks))
  }

  val flat = pending.flatMap { it.chunks }
  val vectors = ArrayList<List<Double>>(flat.size)
  val effectiveBatch = maxOf(1, batchSize)
  var i = 0
  while (i < flat.size) {
    val batch = flat.subList(i, minOf(i + effectiveBatch, flat.size))
    vectors.addAll(embed(batch.map { it.text }))
    i += effectiveBatch
    onProgress?.invoke(minOf(i, flat.size), flat.size)
  }

  var cursor = 0
  var indexed = 0
  for (entry in pending) {
    val indexedChunks = entry.chunks.map { chunk ->
      IndexedChunk(
        id = "${chunk.path}:${chunk.startLine}-${chunk.endLine}",
        path = chunk.path,
        startLine = chunk.startLine,
        endLine = chunk.endLine,
        text = chunk.text,
        vector = vectors[cursor++],
      )
    }
    index.upsertFile(entry.path, entry.hash, indexedChunks)
    indexed += 1
  }

  return IndexUpdateResult(indexed, skipped, removed)
}

/**
 * Embed [query] and return the most semantically similar chunks. Returns nothing
 * for a blank query or an empty index — callers fall back to grep/glob.
 */
fun retrieveFromWorkspace(
  index: WorkspaceIndex,
  embed: (List<String>) -> List<List<Double>>,
  query: String,
  k: Int = DEFAULT_RETRIEVAL_K,
): List<RetrievedChunk> {
  if (query.isBlank() || index.chunkCount == 0) return emptyList()
  val queryVector = embed(listOf(query)).firstOrNull() ?: return emptyList()
  return index.search(queryVector, maxOf(1, k))
}

/** True when the user's message opts into workspace retrieval via `@workspace`. */
fun mentionsWorkspace(text: String): Boolean = Regex("(^|\\s)@workspace\\b", RegexOption.IGNORE_CASE).containsMatchIn(text)
